File: gas_station_problem/python/bucket_solution.py
import sys
import json
import time

MILEAGE = 50
END_MILE = 10000

''' n solution '''

def load_list(input_path):

    with open(input_path) as in_f:
        return json.load(in_f)


def gen_buckets(mileage_list):
    buckets = {}
    for station in mileage_list:
        dist = station // MILEAGE
        if dist in buckets:
            buckets[dist].append(station)
        else:
            buckets[dist] = [station]

    return buckets


def get_gas_stations(mileage_list):

    buckets = gen_buckets(mileage_list)

    output_route = []
    curr_distance = 0
    curr_bucket = 0
    while curr_distance != END_MILE:
        max_station = curr_distance
        for station in buckets[curr_bucket]:
            if station < curr_distance + MILEAGE and station > max_station:
                max_station = station

        if max_station == curr_distance:
            for station in buckets[curr_bucket - 1]:
                if station > max_station:
                    max_station = station
        else:
            curr_bucket += 1

        curr_distance = max_station
        output_route.append(max_station)

    return output_route


def main():
    if len(sys.argv) != 2:
        print('Incorrect number of args, requires input file')
        exit()

    input_list = load_list(sys.argv[1])

    start = time.time()
    output_list = get_gas_stations(input_list)
    end = time.time()
    print(output_list)
    print('Solution took:')
    print(end - start)


if __name__ == '__main__':
    main()

File: gas_station_problem/java/src/BucketSolution.java
import org.json.JSONArray;

import java.util.ArrayList;
import java.util.HashMap;

public class BucketSolution {

    public static ArrayList<Integer> getGasStations(JSONArray inputArr) {
        HashMap<Integer, ArrayList<Integer>> buckets = genBuckets(inputArr);
        ArrayList<Integer> outputArr = new ArrayList<>();

        int currentDistance = 0;
        int currentBucket = 0;
        while (currentDistance != Constants.END_MILE) {
            int maxStation = currentDistance;
            for (Integer station : buckets.get(currentBucket)) {
                if (station < currentDistance + Constants.MILEAGE && station > maxStation) {
                    maxStation = station;
                }
            }

            if (maxStation == currentDistance) {
                for (Integer station : buckets.get(currentBucket - 1)) {
                    if (station > maxStation) {
                        maxStation = station;
                    }
                }
            } else {
                currentBucket += 1;
            }

            currentDistance = maxStation;
            outputArr.add(maxStation);
        }

        return outputArr;
    }

    private static HashMap<Integer, ArrayList<Integer>> genBuckets(JSONArray inputArray) {
        HashMap<Integer, ArrayList<Integer>> buckets = new HashMap<>();
        for (int i = 0; i < inputArray.length(); i++) {
            int station = inputArray.optInt(i);
            int distance = station / Constants.MILEAGE;
            if (!buckets.containsKey(distance)) {
                buckets.put(distance, new ArrayList<>());
            }
            buckets.get(distance).add(station);
        }
        return buckets;
    }
}
